### DISCUSSION TITLE
Various suggestions

### DISCUSSION THREAD

**Discussion by Tellicious** [RavenSystem/esp-homekit-devices#2117] [New Features]
While making some tests with HAA, I came across some ideas for improvement:
- Create a logic in air quality monitor so that the overall AQ value is the highest values received when commanded by multiple FM. E.g.: I have 1 FM reading PM2.5 and 1 FM reading TVOC; if the first one sets AQ to 4, and the second to 3, the overall value should be 4, regardless of what service set it first
- Enable CO2 sensor to show current value in the Home app as part of the accessory (available in some commercial sensors)
- Make setting `"h":2` (allowing a service to be visible only in HAA Manager) to work even if an accessory is not declared as an extra service. Currently if it is not declared as an extra-service, it still appears in Home App
- Implement possibility to enter a deep-sleep mode allowing HAA to be used in battery-powered devices. ESP8266 is currently not suitable for battery-powered devices due to its high consumption (a temperature sensor with DHT22 powered by 4 AAA batteries last 8-12 hours, even with a 15m update time)

**Comment by RavenSystem** [comment]
Thanks for your suggestions.

> Create a logic in air quality monitor so that the overall AQ value is the highest values received when commanded by multiple FM. E.g.: I have 1 FM reading PM2.5 and 1 FM reading TVOC; if the first one sets AQ to 4, and the second to 3, the overall value should be 4, regardless of what service set it first

This is done using an additional Free Monitor Service with Math type, comparing both values, and copying characteristic from highest value to AQ service.

> Enable CO2 sensor to show current value in the Home app as part of the accessory (available in some commercial sensors)

I must check out how it is declared in official Apple HomeKit specs to see if it is possible.

> Make setting "h":2 (allowing a service to be visible only in HAA Manager) to work even if an accessory is not declared as an extra service. Currently if it is not declared as an extra-service, it still appears in Home App

`"h":2` only adds the `hidden` HomeKit flag to selected service. Its behavior is outside my control. From official Apple HomeKit specs, only additional services into an accessory can be set as hidden. You can use `"h":2` as part of accessory, applying hidden flag to main service, but you will find that service is shown as incompatible in Apple Home App. 

> Implement possibility to enter a deep-sleep mode allowing HAA to be used in battery-powered devices. ESP8266 is currently not suitable for battery-powered devices due to its high consumption (a temperature sensor with DHT22 powered by 4 AAA batteries last 8-12 hours, even with a 15m update time)

It is not possible with HAP (HomeKit Accessory Protocol) over WiFi. HAP requires a persistent network connection from controller (HomePods, iPhones, AppleTVs, iPads...). Using any sleep type will disconnect device from WiFi. From official Apple HomeKit specs, to use battery powered devices, Bluetooth LE or Thread is required. HAP over BLE or Thread allow to put device in a dormant state, but it is not possible with HAP over WiFi.

**Comment by Tellicious** [comment]
As always, thanks for the kind reply

Just a couple of questions:

> This is done using an additional Free Monitor Service with Math type, comparing both values, and copying characteristic from highest value to AQ service.

Is there any easier way to do so than calculating `max(a,b)` as `(a+b)/2 + abs(a-b)/2`? I don't see a grater/smaller operator in FM Math wiki

> I must check out how it is declared in official Apple HomeKit specs to see if it is possible.

If It might help, I think there should be dedicated characteristics for both CO2 and CO values i.e. `CarbonDioxideLevel` and `CarbonMonoxideLevel`

**Reply by RavenSystem** [reply]
> Is there any easier way to do so than calculating max(a,b) as (a+b)/2 + abs(a-b)/2? I don't see a grater/smaller operator in FM Math wiki

I don't understand. `max(a,b)` and `(a+b)/2 + abs(a-b)/2` are different things. The first is very simple to do, the second is more complex and probably requires to use 2 FMs Math.

> If It might help, I think there should be dedicated characteristics for both CO2 and CO values i.e. CarbonDioxideLevel and CarbonMonoxideLevel

Yes, I check it out, and CO and CO2 sensors have read-only characteristics for level and peak. I will add an option to enable one or both. Their values must be set using Free Monitor.

**Reply by Tellicious** [reply]
> Yes, I check it out, and CO and CO2 sensors have read-only characteristics for level and peak. I will add an option to enable one or both. Their values must be set using Free Monitor.

Many thanks for that!

> I don't understand. max(a,b) and (a+b)/2 + abs(a-b)/2 are different things. The first is very simple to do, the second is more complex and probably requires to use 2 FMs Math.

Actually  `(a+b)/2 + abs(a-b)/2` is a bad way to calculate max(a,b) using only math operations: if a>b if becomes equivalent to `(a+b)/2 + (a-b)/2 = a` , otherwise if b>a it becomes  `(a+b)/2 + (b-a)/2 = b`. How can I do it with the available operations in Free Monitor? I am not able to find any "greater than" or "smaller than" operator in the wiki

**Reply by RavenSystem** [reply]
It is simple. Only do `a - b` ; If `a > b`, result will be `> 0`; if `a < b`, result will be `< 0`. Use 2 wildcard actions to select to use value from `a` for `"v":0` , and from `b` for `"v":-100000` (less than 0). If `a == b`, result will be `0`, and wildcard action with `"v":0` will be used in this case, but both are right because `a` and `b` are the same value.

**Reply by Tellicious** [reply]
You're right...shame on me for not thinking of that

**Reply by Tellicious** [reply]
Actually it's probably a bit more complex than that. I'm thinking of implementing it as such (with 6 services):

1. Free monitor for Sensor 1 (reads value in PPM) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 1 (hidden)
2. Free monitor for  Sensor 2 (reads value in PPM) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 2 (hidden)
3. Free Monitor Math (comparing data from AQ1 and AQ2) -- wildcard action + service notification (Air Quality value 1-6) --> **Air Quality service 3** (visible)

Having 3 sensors would make thinks even more complex (10 services):

1. Free monitor for Sensor 1 (reads value in PPM) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 1 (hidden)
2. Free monitor for Sensor 2 (reads value in PPM) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 2 (hidden)
3. Free monitor for Sensor 2 (reads value in PPM) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 3 (hidden)
4. Free Monitor Math (comparing data from AQ1 and AQ2) -- wildcard action + service notification (Air Quality value 1-6) --> Air Quality service 4 (hidden)
5. Free Monitor Math (comparing data from AQ3 and AQ4) -- wildcard action + service notification (Air Quality value 1-6) --> **Air Quality service 5** (visible)

Is that correct?

**Reply by RavenSystem** [reply]
Yes. But I recommend to disable HomeKit with `"h":0` in those services that are `hidden`, to save resources. Remember to use `"cd":0.01` or a greater value on these services to avoid concurrence.

**Comment by RavenSystem** [comment]
About enabling CO2 sensor to show current value in the Home app as part of the accessory, try https://github.com/RavenSystem/haabeta/releases/tag/12.9.0b1

**Reply by Tellicious** [reply]
Thanks a lot for the update

I updated to 12.9.0b1 and changed the CO2 sensor service to `{"t":10,"dt":1,"i":60}`, adding `"tg":[n,1]` (where n is the service n) to the related FM, but I'm not getting the CO2 level characteristic visible neither in the Home app nor in HAA Manager (version 3.6.0). Am I doing something wrong?

**Reply by RavenSystem** [reply]
Try new version, please: https://github.com/RavenSystem/haabeta/releases/tag/12.9.0b2

**Reply by Tellicious** [reply]
Tested previous configuration and now it works perfectly. Thanks!!

**Comment by Tellicious** [comment]
Another suggestion that came to my mind is allowing the possibility of reading multiple values with the same I2C FM service. Several sensors are providing multiple values through the same read command, and it would be nice to be able to read them all at once, without using multiple service, with a syntax such as:
   ```
 {
      "t":80,
      "n":21,
      "tg":[[2,1],[2,2],[2,3]],    
      "ic":[0,98],
      "in":[[0,33,177]],
      "it":[0.01,0,236,5],
      "io":[1,1,1],
      "dt":[[2,0],[2,0],[2,0]],
      "l":[[0,1000],[0,500],[400,5000]],
      "y0":[{"v":0,"0":{"m":[[12,30001]]}},
          {"v":50,"r":1,"0":{"m":[[2,1]]}},
          {"v":100,"r":1,"0":{"m":[[2,1]]}}],
      "y1":[{"v":0,"0":{"m":[[12,30001]]}},
          {"v":100,"r":1,"0":{"m":[[2,1],[2,-1],[12,30003]]}},
          {"v":200,"r":1,"0":{"m":[[2,1],[2,-1],[12,30005]]}}]
    }
```

In the example above, with a configuration such as this, you might read 3 values, from a string such as `1B - val 1 (2B) - 1B - val 2 (2B) - 1B - val 3 (2B)`.

I know it's not needed, as it already possible to get all the readings with the current structure but, apart from simplifying the configuration, this would allow to significantly reduce tasks synchronization complexity when it's needed to read multiple values from the same sensor, and also to reduce the amount of read tasks required to the sensor itself

**Reply by RavenSystem** [reply]
It is very complex to manage. A new data structure is needed, and a complete rework on math operations and dynamic HomeKit characteristics. Many breaking changes, and a lot of work, to gain only a shorter script, but using more DRAM and CPU.

**Reply by YouCanNotBeSerious** [reply]
Regarding this:

>Implement possibility to enter a deep-sleep mode allowing HAA to be used in battery-powered devices.

>> It is not possible with HAP (HomeKit Accessory Protocol) over WiFi. HAP requires a persistent network connection ......... to use battery powered devices, Bluetooth LE or Thread is required. 

I was looking for something like this explanation, could that text be added to the wiki to be easily found?  I see at least one other has been in my situation [here.](https://github.com/RavenSystem/esp-homekit-devices/discussions/1766)

Awesome project, I have just started here with it and have an ESP-01S as a simple switch showing up and working in my Home within 1 hour of finding this project :-)
